### PR TITLE
Add the new capability for MRI-4DVar

### PR DESCRIPTION
**Pull request by liujake**
This large PR adds the new capability for multi-resolution incremental 4DVar.

TYPE: new feature

KEYWORDS: MRI-4DVar, Inverse transform, square root of B, 3-stage

SOURCE: Zhiquan (Jake) Liu, Junmei Ban, Jamie Bresch, Yali Wu, Xin Zhang, Jianyu Liu (NCAR/MMM)

DESCRIPTION OF CHANGES:
Existing 4DVar algorithm is run at the same model resolution for inner loop minimization of different outer loops (default namelist setting with multi_inc=0). This adds MRI-4DVar capability that can run 4DVar at lower resolution for inner loop minimization with a 3-stage procedure (observer, minimizer, regridder) as described by Liu et al. (2020) and can substantially speed up 4DVar. Three major changes include:

1) 3-stage procedure controlled by a script and offline programs under ~var/mri4dvar;
Set multi_inc=1 for the "observer" step and multi_inc=2 for the "minimizer" step. The "regridder' step is done using the offline programs.

2) Inverse transform of square root B introduced under ~var/da/da_vtox_transforms;
New namelist parameter use_inverse_squarerootb needs to be set .true. when multi_inc=2 and works together with cv_options=5, 6, or 7 and cloud_cv_options=2 or 3.

New namelist parameter use_interpolate_cvt=.false. (default), which is supposed to work together with the obsolete cv_options=3, which is not recommended to use for non-expert users.

3) Additional file I/O for control variable and radar/radiance OMB files from the observer step. This I/O part will need some further code clean-up and improvement in the future.

LIST OF MODIFIED FILES: 48 files
M       Registry/registry.var
M       var/build/depend.txt
M       var/da/da_main/da_solve.inc
M       var/da/da_main/da_wrfvar_top.f90
M       var/da/da_minimisation/da_get_innov_vector.inc
M       var/da/da_minimisation/da_minimisation.f90
M       var/da/da_obs_io/da_obs_io.f90
M       var/da/da_obs_io/da_read_iv_for_multi_inc.inc
M       var/da/da_obs_io/da_read_obs_radar.inc
M       var/da/da_obs_io/da_scan_obs_radar.inc
M       var/da/da_obs_io/da_search_obs.inc
M       var/da/da_obs_io/da_write_iv_for_multi_inc.inc
M       var/da/da_radiance/da_radiance1.f90
A       var/da/da_radiance/da_read_iv_rad_for_multi_inc.inc
A       var/da/da_radiance/da_write_iv_rad_for_multi_inc.inc
M       var/da/da_recursive_filter/da_recursive_filter.f90
A       var/da/da_recursive_filter/da_recursive_filter_1d_inv.inc
M       var/da/da_recursive_filter/da_transform_through_rf.inc
A       var/da/da_recursive_filter/da_transform_through_rf_inv.inc
M       var/da/da_setup_structures/da_setup_structures.f90
A       var/da/da_setup_structures/da_write_vp.inc
M       var/da/da_tools/da_llxy.inc
M       var/da/da_vtox_transforms/da_transform_vptox.inc
A       var/da/da_vtox_transforms/da_transform_vptox_inv.inc
A       var/da/da_vtox_transforms/da_transform_vtovv_inv.inc
A       var/da/da_vtox_transforms/da_transform_vtox_inv.inc
M       var/da/da_vtox_transforms/da_transform_vvtovp.inc
M       var/da/da_vtox_transforms/da_transform_vvtovp_adj.inc
A       var/da/da_vtox_transforms/da_transform_vvtovp_inv.inc
M       var/da/da_vtox_transforms/da_vertical_transform.inc
M       var/da/da_vtox_transforms/da_vtox_transforms.f90
A       var/mri4dvar/Makefile
A       var/mri4dvar/README.MRI-4DVar
A       var/mri4dvar/da_bdy.f90
A       var/mri4dvar/da_bilin.f90
A       var/mri4dvar/da_thin.f90
A       var/mri4dvar/da_vp_bilin.f90
A       var/mri4dvar/da_vp_split.f90
A       var/mri4dvar/nc_increment.ncl
A       var/mri4dvar/nc_vpglobal.ncl
A       var/mri4dvar/nc_vphires.ncl
A       var/mri4dvar/nc_vpinput.ncl
A       var/mri4dvar/nc_vplocal.ncl
A       var/mri4dvar/rsl_lite.h
A       var/mri4dvar/run_mri3d4dvar.csh_lsf
A       var/mri4dvar/run_mri3d4dvar.csh_pbs
A       var/mri4dvar/task_for_point.c
A       var/mri4dvar/wraper_mri3d4dvar.csh

TESTS CONDUCTED: WRFDA regression test passed. Cycling MRI-4DVar tests were performed over an Arctic domain.

RELEASE NOTE: New capability for Multi-Resolution Incremental 4DVar (MRI-4DVar).
Liu, Z., J. Ban, J.-S, Hong, and Y.-H. Kuo, 2020: Multi-resolution incremental 4D-Var for WRF: Implementation and application at convective scale, Q. J. R. Meteorol. Soc., 146, 3661-3674.

### CODE REVIEW

**Comment by liujake** [comment]
@jamiebresch I see da_read/write_iv_for_multi_inc_opt2.inc only differs from da_read/write_iv_for_multi_inc.inc in radar I/O part. If you have no objection, I plan to replace the old code with your new code instead of keeping both.

File: var/da/da_obs_io/da_read_obs_radar.inc [hunk 1]
```diff
@@ -219,7 +219,7 @@ subroutine da_read_obs_radar (iv, filename, grid)
          endif
 
          call da_llxy (platform%info, platform%loc, outside, outside_all)
-         if( outside_all ) then
+         if( outside_all .and. multi_inc == 0 ) then
```
Comment (liujake) [inline var/da/da_obs_io/da_read_obs_radar.inc:222]: @jamiebresch You or I added this one line change? No same change for other obs types. Looks like other obs types do not write out QC and obs error info, but your more efficient binary radar I/O writes out QC and obs error info in observer mode. Is that the reason for this one line change?

**Comment by liujake** [comment]
@davegill Listed you as a reviewer just in case you'd like to have a quick scan in PR message (and/or code) for this relatively large PR. No impact to the model side. WRFDA regtest passed. Junmei should review it.

**Review by junmeiban — APPROVED** [review]
I tested this PR on OSC HPC owens machine. 
Two cycling experiments(135km45km and 45km45km) are conducted, and both of them run successfully.

**Pull request merged**